File: java/graphs/cycle_det_directed.java
import java.util.ArrayList;

public class cycle_det_directed {
    public boolean isCyclic(int V, int[][] edges) {
        ArrayList<ArrayList<Integer>> adj= new ArrayList<>();
        for(int i=0;i<V;i++){
            adj.add(new ArrayList<>());
        }
        for(int i=0;i<edges.length;i++){
            int u=edges[i][0];
            int v=edges[i][1];
            adj.get(u).add(v);
        }
        boolean vis[]= new boolean[V];
        boolean path[]= new boolean[V];
        for(int i=0;i<V;i++){
            if(!vis[i] && isCyc(vis,adj,i,path)){
                return true;
            }
        }
        return false;
    }
    boolean isCyc(boolean vis[],ArrayList<ArrayList<Integer>> adj,int node,boolean path[]){
        vis[node]=true;
        path[node]= true;
        for(int nei:adj.get(node)){
            if(!vis[nei]){
                if(isCyc(vis,adj,nei,path)) return true;
            }
            else if(path[nei]){
                return true;
            }
        }
        path[node]=false;
        return false;
    }
}

File: python/graphs/cycle_det_directed.py
def is_cyclic(V, edges):
    # Build adjacency list
    adj = [[] for _ in range(V)]
    for u, v in edges:
        adj[u].append(v)

    visited = [False] * V
    path = [False] * V

    for i in range(V):
        if not visited[i] and is_cyc(visited, adj, i, path):
            return True
    return False

def is_cyc(visited, adj, node, path):
    visited[node] = True
    path[node] = True

    for nei in adj[node]:
        if not visited[nei]:
            if is_cyc(visited, adj, nei, path):
                return True
        elif path[nei]:
            return True

    path[node] = False
    return False

# Example usage
if __name__ == "__main__":
    V = 4
    edges = [
        [0, 1],
        [1, 2],
        [2, 3],
        [3, 1]  # This creates a cycle
    ]
    print("Graph has cycle:", is_cyclic(V, edges))
